org.graalvm.python.embedding.tools/src/main/java/org/graalvm/python/embedding/tools/vfs/VFSUtils.java

```diff
@@ -539,54 +528,93 @@ public List<String> getLockFilePackages() {
 			throw new IllegalStateException(e);
 		}
 	}
 
-	public static void createVenv(Path venvDirectory, List<String> packages, Path lockFilePath,
-			String missingLockFileWarning, Launcher launcher, String graalPyVersion, BuildToolLog log)
-			throws IOException, PackagesChangedException {
-		Objects.requireNonNull(venvDirectory);
-		Objects.requireNonNull(packages);
-		Objects.requireNonNull(launcher);
-		Objects.requireNonNull(graalPyVersion);
-		Objects.requireNonNull(log);
+    public static void createVenv(Path venvDirectory, List<String> packages, Path lockFilePath,
+            String missingLockFileWarning, Launcher launcher, String graalPyVersion, BuildToolLog log, Path reqFile
+    ) throws IOException, PackagesChangedException {
 
-		logVenvArgs(venvDirectory, packages, lockFilePath, launcher, graalPyVersion, log);
+        Objects.requireNonNull(venvDirectory);
+        Objects.requireNonNull(launcher);
+        Objects.requireNonNull(graalPyVersion);
+        Objects.requireNonNull(log);
 
-		List<String> pluginPackages = trim(packages);
-		LockFile lockFile = null;
-		if (lockFilePath != null && Files.exists(lockFilePath)) {
-			lockFile = LockFile.fromFile(lockFilePath, log);
-		}
+        if (installPackagesFromReqFile(venvDirectory, launcher, graalPyVersion, log, reqFile)) {
+            return;
+        }
 
-		if (!checkPackages(venvDirectory, pluginPackages, lockFile, log)) {
-			return;
-		}
+        installPackages(venvDirectory, packages, lockFilePath, missingLockFileWarning, launcher, graalPyVersion, log);
+    }
 
-		VenvContents venvContents = ensureVenv(venvDirectory, graalPyVersion, launcher, log);
+    static boolean installPackagesFromReqFile(Path venvDirectory, Launcher launcher,
+            String graalPyVersion, BuildToolLog log, Path reqFile) throws IOException {
+        if (reqFile != null) {
+            log.info("Using <requirements.txt> dependency mode.");
+            log.info("Installing Python dependencies from: " + reqFile);
+            log.warning("Lock file is ignored in <requirements.txt> mode.");
+            log.warning("The 'lock-packages' goal should not be used together with <requirementsFile>.");
 
-		InstalledPackages installedPackages = InstalledPackages.fromVenv(venvDirectory);
-		boolean installed;
-		if (lockFile != null) {
-			installed = install(venvDirectory, installedPackages, lockFile, log);
-		} else {
-			installed = install(venvDirectory, pluginPackages, venvContents, log);
-		}
-		if (installed) {
-			venvContents.write(pluginPackages);
-			installedPackages.freeze(log);
-		}
-		if (lockFile == null) {
-			missingLockFileWarning(venvDirectory, pluginPackages, missingLockFileWarning, log);
-		}
-	}
+            ensureVenv(venvDirectory, graalPyVersion, launcher, log);
+
+            runPip(venvDirectory, "install", log, "--compile", "-r", reqFile.toString());
+
+            InstalledPackages installed = InstalledPackages.fromVenv(venvDirectory);
+            installed.freeze(log);
```

Comment:
You need to save `VenvContents vc = ensureVenv(venvDirectory, graalPyVersion, launcher, log)`, so that here you can do `vc.write(requirementsPackages(reqFile));` -- we need to write those meta-data in the venv, so that `ensureVenv` is not going to recreate it on next build.